psci/PSCi.hs

```diff
@@ -376,20 +378,66 @@ handleTypeOf val = do
+           printModule's "data constructors" (map showDataConstructor . sortBy compareDatatypes . map (findDataConstructor dataConstructors)) moduleDataConstructors
+           printModule's "functions"         (sort . map (showNameType . findNameType names)) moduleNamesIdent
+           outputStrLn   ""
+
+
+  where printModule's what _  [] = outputStrLn $ "This module '" ++ P.runModuleName moduleName ++ "' does not export " ++ what ++ ".\n"
+        printModule's _ showF xs = (outputStr . unlines . showF) xs
+
+        findNameType :: M.Map (P.ModuleName, P.Ident) (P.Type, P.NameKind, P.NameVisibility) -> (P.ModuleName, P.Ident) -> (P.Ident, Maybe (P.Type, P.NameKind, P.NameVisibility))
+        findNameType envNames m@(_, mIdent) = (mIdent, M.lookup m envNames)
+
+        showNameType :: (P.Ident, Maybe (P.Type, P.NameKind, P.NameVisibility)) -> String
+        showNameType (mIdent, Just (mType, _, _)) = Box.render (Box.text (P.showIdent mIdent ++ " :: ") Box.<> P.typeAsBox mType)
+        showNameType _ = P.internalError "The impossible happened in printModuleSignatures."
+
+        findDataConstructor :: M.Map (P.Qualified P.ProperName) (P.DataDeclType, P.ProperName, P.Type, [P.Ident]) -> P.Qualified P.ProperName -> (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident]))
+        findDataConstructor envDataCons name = (name, M.lookup name envDataCons)
+
+        compareDatatypes :: (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident])) -> (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident])) -> Ordering
+        compareDatatypes (_, Just (_, dtName1, _, _)) (_, Just (_, dtName2, _, _)) = dtName1 `compare` dtName2
+        compareDatatypes _ _ = P.internalError "The impossible happened in printModuleSignatures."
+
+        showDataConstructor :: (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident])) -> String
+        showDataConstructor (P.Qualified _ name, Just (_, _, dtType, _)) = Box.render (Box.text (P.runProperName name ++ " :: ") Box.<> P.typeAsBox dtType)
+        showDataConstructor _ = P.internalError "The impossible happened in printModuleSignatures."
+
+        findType :: M.Map (P.Qualified P.ProperName) (P.Kind, P.TypeKind) -> P.Qualified P.ProperName -> (P.Qualified P.ProperName, Maybe (P.Kind, P.TypeKind))
+        findType envTypes name = (name, M.lookup name envTypes)
+
+        showType :: M.Map (P.Qualified P.ProperName) ([(String, Maybe P.Kind)], P.Type) -> (P.Qualified P.ProperName, Maybe (P.Kind, P.TypeKind)) -> Maybe String
+        showType typeSynonymsEnv (n@(P.Qualified _ name), typ) =
+          case (typ, M.lookup n typeSynonymsEnv) of
+            (Just (_, P.TypeSynonym), Just (typevars, dtType)) ->
+                Just (Box.render $ Box.text ("type " ++ P.runProperName name ++ concatMap ((' ':) . fst) typevars) Box.// Box.moveRight 2 (Box.text "=" Box.<+> P.typeAsBox dtType))
```

Comment:
wasn't sure exactly how to use unwords here, did you mean something like `(' ':) . unwords . map fst` ?